File: android/src/main/kotlin/com/pravera/flutter_foreground_task/models/NotificationContent.kt
package com.pravera.flutter_foreground_task.models

import android.content.Context
import com.pravera.flutter_foreground_task.PreferencesKey as PrefsKey
import org.json.JSONArray
import org.json.JSONObject

data class NotificationContent(
        val title: String,
        val text: String,
        val icon: NotificationIcon?,
        val buttons: List<NotificationButton>,
        val initialRoute: String?
) {
    companion object {
        fun getData(context: Context): NotificationContent {
            val prefs = context.getSharedPreferences(
                PrefsKey.NOTIFICATION_OPTIONS_PREFS, Context.MODE_PRIVATE)

            val title = prefs.getString(PrefsKey.NOTIFICATION_CONTENT_TITLE, null) ?: ""
            val text = prefs.getString(PrefsKey.NOTIFICATION_CONTENT_TEXT, null) ?: ""

            val iconJsonString = prefs.getString(PrefsKey.NOTIFICATION_CONTENT_ICON, null)
            var icon: NotificationIcon? = null
            if (iconJsonString != null) {
                icon = NotificationIcon.fromJsonString(iconJsonString)
            }

            val buttonsJsonString = prefs.getString(PrefsKey.NOTIFICATION_CONTENT_BUTTONS, null)
            val buttons: MutableList<NotificationButton> = mutableListOf()
            if (buttonsJsonString != null) {
                val buttonsJsonArr = JSONArray(buttonsJsonString)
                for (i in 0 until buttonsJsonArr.length()) {
                    val buttonJsonObj = buttonsJsonArr.getJSONObject(i)
                    buttons.add(NotificationButton.fromJSONObject(buttonJsonObj))
                }
            }

            val initialRoute = prefs.getString(PrefsKey.NOTIFICATION_INITIAL_ROUTE, null)

            return NotificationContent(
                title = title,
                text = text,
                icon = icon,
                buttons = buttons,
                initialRoute = initialRoute
            )
        }

        fun setData(context: Context, map: Map<*, *>?) {
            val prefs = context.getSharedPreferences(
                PrefsKey.NOTIFICATION_OPTIONS_PREFS, Context.MODE_PRIVATE)

            val title = map?.get(PrefsKey.NOTIFICATION_CONTENT_TITLE) as? String ?: ""
            val text = map?.get(PrefsKey.NOTIFICATION_CONTENT_TEXT) as? String ?: ""

            val iconJson = map?.get(PrefsKey.NOTIFICATION_CONTENT_ICON) as? Map<*, *>
            var iconJsonString: String? = null
            if (iconJson != null) {
                iconJsonString = JSONObject(iconJson).toString()
            }

            val buttonsJson = map?.get(PrefsKey.NOTIFICATION_CONTENT_BUTTONS) as? List<*>
            var buttonsJsonString: String? = null
            if (buttonsJson != null) {
                buttonsJsonString = JSONArray(buttonsJson).toString()
            }

            val initialRoute = map?.get(PrefsKey.NOTIFICATION_INITIAL_ROUTE) as? String

            with(prefs.edit()) {
                putString(PrefsKey.NOTIFICATION_CONTENT_TITLE, title)
                putString(PrefsKey.NOTIFICATION_CONTENT_TEXT, text)
                putString(PrefsKey.NOTIFICATION_CONTENT_ICON, iconJsonString)
                putString(PrefsKey.NOTIFICATION_CONTENT_BUTTONS, buttonsJsonString)
                putString(PrefsKey.NOTIFICATION_INITIAL_ROUTE, initialRoute)
                commit()
            }
        }

        fun updateData(context: Context, map: Map<*, *>?) {
            val prefs = context.getSharedPreferences(
                PrefsKey.NOTIFICATION_OPTIONS_PREFS, Context.MODE_PRIVATE)

            val title = map?.get(PrefsKey.NOTIFICATION_CONTENT_TITLE) as? String
            val text = map?.get(PrefsKey.NOTIFICATION_CONTENT_TEXT) as? String

            val iconJson = map?.get(PrefsKey.NOTIFICATION_CONTENT_ICON) as? Map<*, *>
            var iconJsonString: String? = null
            if (iconJson != null) {
                iconJsonString = JSONObject(iconJson).toString()
            }

            val buttonsJson = map?.get(PrefsKey.NOTIFICATION_CONTENT_BUTTONS) as? List<*>
            var buttonsJsonString: String? = null
            if (buttonsJson != null) {
                buttonsJsonString = JSONArray(buttonsJson).toString()
            }

            val initialRoute = map?.get(PrefsKey.NOTIFICATION_INITIAL_ROUTE) as? String

            with(prefs.edit()) {
                title?.let { putString(PrefsKey.NOTIFICATION_CONTENT_TITLE, it) }
                text?.let { putString(PrefsKey.NOTIFICATION_CONTENT_TEXT, it) }
                iconJsonString?.let { putString(PrefsKey.NOTIFICATION_CONTENT_ICON, it) }
                buttonsJsonString?.let { putString(PrefsKey.NOTIFICATION_CONTENT_BUTTONS, it) }
                initialRoute?.let { putString(PrefsKey.NOTIFICATION_INITIAL_ROUTE, it) }
                commit()
            }
        }

        fun clearData(context: Context) {
            val prefs = context.getSharedPreferences(
                PrefsKey.NOTIFICATION_OPTIONS_PREFS, Context.MODE_PRIVATE)

            with(prefs.edit()) {
                clear()
                commit()
            }
        }
    }
}

File: ios/flutter_foreground_task/Sources/flutter_foreground_task/models/NotificationContent.swift
//
//  NotificationContent.swift
//  flutter_foreground_task
//
//  Created by Woo Jin Hwang on 8/2/24.
//

import Foundation

struct NotificationContent {
  let title: String
  let text: String
  let buttons: Array<NotificationButton>
  
  static func getData() -> NotificationContent {
    let prefs = UserDefaults.standard
    
    let title = prefs.string(forKey: NOTIFICATION_CONTENT_TITLE) ?? ""
    let text = prefs.string(forKey: NOTIFICATION_CONTENT_TEXT) ?? ""
    
    var buttons: [NotificationButton] = []
    if let buttonsJsonString = prefs.string(forKey: NOTIFICATION_CONTENT_BUTTONS) {
      if let buttonsJsonData = buttonsJsonString.data(using: .utf8) {
        if let buttonsJsonArr = try? JSONSerialization.jsonObject(with: buttonsJsonData, options: []) as? [Any] {
          for buttonJsonObj in buttonsJsonArr {
            buttons.append(NotificationButton.fromJSONObject(buttonJsonObj))
          }
        }
      }
    }
    
    return NotificationContent(title: title, text: text, buttons: buttons)
  }
  
  static func setData(args: Dictionary<String, Any>) {
    let prefs = UserDefaults.standard
    
    let title = args[NOTIFICATION_CONTENT_TITLE] as? String ?? ""
    prefs.set(title, forKey: NOTIFICATION_CONTENT_TITLE)
    
    let text = args[NOTIFICATION_CONTENT_TEXT] as? String ?? ""
    prefs.set(text, forKey: NOTIFICATION_CONTENT_TEXT)
    
    if let buttonsJson = args[NOTIFICATION_CONTENT_BUTTONS] as? [Any] {
      if let buttonsJsonData = try? JSONSerialization.data(withJSONObject: buttonsJson, options: []) {
        if let buttonsJsonString = String(data: buttonsJsonData, encoding: .utf8) {
          prefs.set(buttonsJsonString, forKey: NOTIFICATION_CONTENT_BUTTONS)
        }
      }
    }
  }
  
  static func updateData(args: Dictionary<String, Any>) {
    let prefs = UserDefaults.standard
    
    if let title = args[NOTIFICATION_CONTENT_TITLE] as? String {
      prefs.set(title, forKey: NOTIFICATION_CONTENT_TITLE)
    }
    
    if let text = args[NOTIFICATION_CONTENT_TEXT] as? String {
      prefs.set(text, forKey: NOTIFICATION_CONTENT_TEXT)
    }
    
    if let buttonsJson = args[NOTIFICATION_CONTENT_BUTTONS] as? [Any] {
      if let buttonsJsonData = try? JSONSerialization.data(withJSONObject: buttonsJson, options: []) {
        if let buttonsJsonString = String(data: buttonsJsonData, encoding: .utf8) {
          prefs.set(buttonsJsonString, forKey: NOTIFICATION_CONTENT_BUTTONS)
        }
      }
    }
  }
  
  static func clearData() {
    let prefs = UserDefaults.standard
    prefs.removeObject(forKey: NOTIFICATION_CONTENT_TITLE)
    prefs.removeObject(forKey: NOTIFICATION_CONTENT_TEXT)
    prefs.removeObject(forKey: NOTIFICATION_CONTENT_BUTTONS)
  }
}
